File: Javascript/21.merge-two-sorted-lists.js
/*
 * @lc app=leetcode id=21 lang=javascript
 *
 * [21] Merge Two Sorted Lists
 */

/**
 * 
 * @param {number} val 
 * @param {ListNode} next 
 */
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    const initialNode = new ListNode(0);
    let current = initialNode;
    while (list1 !== null || list2 !== null) {
        if (list1 === null) {
            current.next = new ListNode(list2.val);
            current = current.next;
            list2 = list2.next;
            continue;
        }

        if (list2 === null) {
            current.next = new ListNode(list1.val);
            current = current.next;
            list1 = list1.next;
            continue;
        }

        if (list1.val < list2.val) {
            current.next = new ListNode(list1.val);
            current = current.next;
            list1 = list1.next;
            continue;
        }

        if (list2 !== null) {
            current.next = new ListNode(list2.val);
            current = current.next;
            list2 = list2.next;
            continue;
        }
    }

    return initialNode.next;
};
// @lc code=end

File: Ruby/21.merge-two-sorted-lists.rb
#
# @lc app=leetcode id=21 lang=ruby
#
# [21] Merge Two Sorted Lists
#

# @lc code=start
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end
# @param {ListNode} list1
# @param {ListNode} list2
# @return {ListNode}
def merge_two_lists(list1, list2)
    head = ListNode.new
    current = head
    current1 = list1
    current2 = list2

    while current1 != nil || current2 != nil
        if current1 == nil
            current.next = current2
            current2 = current2.next
        elsif current2 == nil
            current.next = current1
            current1 = current1.next
        elsif current2.val < current1.val
            current.next = current2
            current2 = current2.next
        else current1.val < current2.val
            current.next = current1
            current1 = current1.next
        end

        current = current.next
    end

    return head.next
end
# @lc code=end

merge_two_lists(ListNode.new(1, ListNode.new(2, ListNode.new(4))), ListNode.new(1, ListNode.new(3, ListNode.new(4))))